src/librbd/api/Mirror.cc

```diff
@@ -2098,6 +2345,1689 @@ void Mirror<I>::image_snapshot_create(I *ictx, uint32_t flags,
+  }
+}
+
+template <typename I>
+int is_images_mirror_enabled(std::vector<I *> *image_ctxs) {
+  std::vector<mirror_image_info_t> mirror_images(image_ctxs->size());
+  std::vector<C_SaferCond> on_finishes(image_ctxs->size());
+  int ret_code = 0;
+  ImageCtx *ictx;
+  for (size_t i = 0; i < image_ctxs->size(); i++) {
+    ictx = (*image_ctxs)[i];
+    librbd::api::Mirror<>::image_get_info(ictx,
+                                          &mirror_images[i], &on_finishes[i]);
+  }
+
+  for (size_t i = 0; i < image_ctxs->size(); i++) {
+    int r = on_finishes[i].wait();
+    if (r < 0) {
+      if (ret_code == 0) {
+        ret_code = r;
+      }
+    } else if (mirror_images[i].state == RBD_MIRROR_IMAGE_ENABLED) {
+      ictx = (*image_ctxs)[i];
+      lderr(ictx->cct) << "image " << ictx->name
+                       << " has mirroring enabled" << dendl;
+      ret_code = -EINVAL;
+      break;
+    }
+  }
+
+  return ret_code;
+}
+
+template <typename I>
+int Mirror<I>::group_list(IoCtx& io_ctx, std::vector<std::string> *names) {
+  CephContext *cct = reinterpret_cast<CephContext *>(io_ctx.cct());
+  ldout(cct, 20) << dendl;
+
+  std::set<std::string> group_ids;
+  std::string last_read = "";
+  int max_read = 1024;
+  int r;
+  do {
+    std::map<std::string, cls::rbd::MirrorGroup> mirror_groups;
+    r =  cls_client::mirror_group_list(&io_ctx, last_read, max_read,
+                                       &mirror_groups);
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "error listing mirrored image directory: "
+                 << cpp_strerror(r) << dendl;
+      return r;
+    }
+    for (auto &[group_id, mirror_group] : mirror_groups) {
+      group_ids.insert(group_id);
+    }
+    if (!mirror_groups.empty()) {
+      last_read = mirror_groups.rbegin()->first;
+    }
+    r = mirror_groups.size();
+  } while (r == max_read);
+
+  if (group_ids.empty()) {
+    return 0;
+  }
+
+  std::map <std::string, std::string> name_to_id_map;
+  r = Group<I>::list(io_ctx, &name_to_id_map);
+  if (r < 0) {
+    return r;
+  }
+
+  for (auto &[name, group_id] : name_to_id_map) {
+    if (group_ids.count(group_id) > 0) {
+      names->push_back(name);
+    }
+  }
+  return 0;
+}
+
+template <typename I>
+int Mirror<I>::group_enable(IoCtx& group_ioctx, const char *group_name,
+                            mirror_image_mode_t mirror_image_mode,
+                            uint32_t flags) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  ldout(cct, 20) << "io_ctx=" << &group_ioctx
+		 << ", group_name=" << group_name
+		 << ", namespace=" << group_ioctx.get_namespace()
+		 << ", mirror_image_mode=" << mirror_image_mode << dendl;
+
+  if (mirror_image_mode != RBD_MIRROR_IMAGE_MODE_SNAPSHOT) {
+    return -EOPNOTSUPP;
+  }
+
+  std::string group_id;
+  int r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
+				 group_name, &group_id);
+  if (r < 0) {
+    lderr(cct) << "error reading group id object: " << cpp_strerror(r) << dendl;
```

Comment:
```suggestion
    lderr(cct) << "error getting the group id : " << cpp_strerror(r) << dendl;
```

There is no group id object. This is reading the RBD_GROUP_DIRECTORY object.